File: Algorithm/Python/palindrome_linked_list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    # @param {ListNode} head
    # @return {boolean}
    def isPalindrome(self, head):
        lst = []
        while head:
            lst.append(head.val)
            head = head.next
        right = len(lst)-1
        left = 0
        while left < right:
            if lst[left] == lst[right]:
                left += 1
                right -= 1
            else:
                return False
        return True 



if __name__ == '__main__':
    sol = Solution()
    import random
    lst = [random.randint(1,100) for i in range(5)]
    print lst
    p = head = ListNode(-1)
    for i in lst:
        p.next = ListNode(i)
        p = p.next
    p.next = None
    print sol.isPalindrome(head.next)

File: Algorithm/CPP/palindrome_linked_list.cpp
# include <vector>
# include <algorithm>
# include <string>
# include <stack>
# include <algorithm>
# include <deque>
# include <iostream>
# include <time.h>
using namespace std;

struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	bool isPalindrome(ListNode* head)
	{
		if (!head) return true;
		auto slow = head;
		auto fast = head;
		stack<int> Stack;
		while (fast && fast->next)
		{
			Stack.push(slow->val);
			slow = slow->next;
			fast = fast->next->next;
			
		}
		if (fast) slow = slow->next;// the odd situation
		while (slow)
		{			
			if (Stack.top() != slow->val) return false;
			Stack.pop();
			slow = slow->next;
		}
		return true;
	}
};
int main()
{
	ListNode *head1 = new ListNode(1);
	ListNode *head2 = new ListNode(-2);
	auto p = head1;
	auto q = head2;
	vector<int> vec{ 2,2,1};//2
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;
	vec = { 3, 4, 7 };
	for (auto v : vec)
	{
		q->next = new ListNode(v);
		q = q->next;
	}
	q->next = NULL;

	Solution sol;

	cout<< sol.isPalindrome(head1);
	cout << endl;
	system("pause");
	return 0;
}
